### DISCUSSION TITLE
Vue 3: Props are visible using the element inspector

### DISCUSSION THREAD

**Discussion by dubljanin96** [inertiajs/inertia#372] [Help (Vue)] [answered]
I have this problem with Inertiajs and Vue3 where props are visible when I do inspect element. Is it okay? 
How to change this? Have I made any mistakes?
<img width="1280" alt="Inertia js" src="https://user-images.githubusercontent.com/29163556/103178235-4c1ecc80-4881-11eb-80c4-60bbf92609f5.png">

**Comment by reinink** [comment, marked as answer]
This is totally expected behaviour. All data that you pass from your server-side framework to Inertia is publicly available.

As noted [from the docs](https://inertiajs.com/responses):

> To ensure that pages load quickly, only return the minimum data required for the page. Also, be aware that all data returned from the controllers will be visible client-side, so be sure to omit sensitive information.

**Reply by muuvmuuv** [reply]
I know this is old but would it be possible to store it in a head meta tag instead of the app div? The `data-page` is so long, I cannot navigate in the Elements view of chrome debugger properly.

**Comment by dubljanin96** [comment]
Why is not same thing happening with with Vue 2?
![image](https://user-images.githubusercontent.com/29163556/103468447-4fbac380-4d59-11eb-85ef-5be579e50aa0.png)

**Reply by grantholle** [reply]
It does happen with vue 2 when the page first loads, but it removes the data attribute from the element.

**Reply by claudiodekker** [reply]
In Vue2, the entire element that you're mounting on gets consumed and replaced: https://vuejs.org/v2/api/#el
> The provided element merely serves as a mounting point. Unlike in Vue 1.x, the mounted element will be replaced with Vue-generated DOM in all cases. It is therefore not recommended to mount the root instance to <html> or <body>.

**Comment by ajnsn** [comment]
@BlueSilver001 

Still not completly hiding it but removing it after the app is mounted, you may try out the following. 
Notice the `onMounted` hook.

```js
import { createApp, h, onMounted } from 'vue';
import { App, plugin } from '@inertiajs/inertia-vue3';

const el = document.getElementById('app');

createApp({
    setup() {
        onMounted(() => {
            delete el.dataset.page;
        });
    },
    render: () =>
        h(App, {
            initialPage: JSON.parse(el.dataset.page),
            resolveComponent: name => require(`./Pages/${name}`).default,
        }),
})
    .use(plugin)
    .mount(el);
```

**Reply by claudiodekker** [reply]
Do note that while doing this visually _hides_ the data from inspection, it is still there. Even if you have the very best tools and techniques available to hide this information, **any type of confidential information (such as passwords) that is sent to the client's browser _should not ever be trusted_**. Simply the fact that the browser has _received_ this data, means that as long as the end user is _smart_ enough and uses the right tools (e.g. a network proxy, memory dump and hex editor, etc.), they _can still_ extract information regardless.

**Reply by ajnsn** [reply]
It's pure cosmetic, yes. Works pretty well for this purpose 👍

**Reply by stevebauman** [reply]
I like this solution -- I regularly modify the dom while playing with Tailwind classes, so removing it after the page has mounted really helps clear up the clutter and I can navigate to things much faster. Thanks @ajnsn!

**Reply by reinink** [reply]
Pretty interesting solution. ALMOST makes me want to do this by default. 🤔 

Obviously, as @claudiodekker notes, this isn't a security feature in any way. But it would be nice to just remove the clutter.

**Reply by stevebauman** [reply]
@reinink Yea I think this would be a great feature by default! I’ve never personally needed to inspect the `data-page` JSON, and if I did I could just view source.

I don’t really see any downsides, besides people possibly making the assumption that the shared data is private (though you have plenty of warnings on the docs for that anyway).

Maybe at least a mention on the docs for anyone else looking for this as well? I can PR if you like!

**Reply by grantholle** [reply]
I find myself adding this:

```js
const cleanApp = () => {
  document.getElementById('app').removeAttribute('data-page')
}

document.addEventListener('inertia:finish', cleanApp)
```

and calling...

```js
createInertiaApp({
  //
}).then(cleanApp)
```

**Reply by aneeskhan47** [reply]
@reinink it would be great if inertia-vue3 do this by default

**Reply by claudiodekker** [reply]
We could consider this as an idea that could be worked on, but it's difficult to track interest here. As such:

- If there's already an idea for this in the ideas section, please vote it up.
- If there's no idea for this yet in the ideas section, please create one for this.

Thanks!

**Reply by Semvrij** [reply]
Would it be possible to remove the `data-page` attribute completely when using SSR?

**Reply by insoutt** [reply]
I implemented a solution to remove `data-page` in SSR with React (would be the same with Vue), edit `ssr.js` to modify the body of the request when `createInertiaApp` is resolved.

```js
createServer((page) => createInertiaApp({
    page,
    render: ReactDOMServer.renderToString,
    resolve: name => require(`./Pages/${name}`),
    setup: ({ App, props }) => <App {...props} />,
}).then((app) => {
    app.body = app.body.replace(/data-page=".*?"/, '')
    return app;
}))
```

Use with caution, I do not know if is recommended to remove `data-page` in SSR.

**Reply by anotherglitchinthematrix** [reply]
Also does the job, with a one-liner.

```javascript
import { router } from '@inertiajs/vue2';
router.on('navigate', () => document.getElementById('app').removeAttribute('data-page'));
```

**Reply by CamKem** [reply]
Also does the job too:
```
.mount(el)
.$nextTick(() => { delete el.dataset.page })
```

**Reply by MohammadZarifiyan** [reply]
This solution is not working anymore.

**Reply by MohammadZarifiyan** [reply]
I need to remove `data-page` attribute from my application, Sometime this attribute value makes my web page 2~4 MB larger.

**Reply by stevebauman** [reply]
@MohammadZarifiyan, removing the `data-page` attribute is only cosmetic and is done after Inertia starts on the client. Inertia requires this property.

If your `data-page` attribute is large, then that means the data you're sending to Inertia is that large.

**Reply by MohammadZarifiyan** [reply]
@stevebauman 
I think there must be a way to remove this attribute after rendering page with SSR.
For example:
1 - Inertia must render the page using `data-page` value that coming from Laravel.
2 - Then it removes the `data-page` attribute from final HTML result.
3 - Sends final result to browser.

But if Inertia JS still needs `data-page` attribute value, SSR must store value somewhere in database and give and ID to client that refers to data stored in database.
After rendering page in browser Inertia must fetch `data-page` data from SSR database using a simple HTTP Request before calling Vue `beforeCreate()` hook. That API call should not block rendering page in browser.

By this approach page will load faster and Inertia still can have access to `data-page` attribute value in client side.

**Comment by GeorgeFourkas** [comment]
You can basically do this:

```js
createInertiaApp({
    title: (title) => `${title} - ${appName}`,
    resolve: (name) => resolvePageComponent(`./Pages/${name}.vue`, import.meta.glob('./Pages/**/*.vue')),
    setup({el, App, props, plugin}) {
    
        const application = createApp({render: () => h(App, props)})
            .use(plugin)
            .use(ZiggyVue, Ziggy)
            .mount(el);
            
        delete el.dataset.page

        return application;
    },
    progress: {
        color: '#4B5563',
    },
 ```
    
So what we've done here is save the creation of the vue Instance into a variable called application, then after the data-page have been parsed we delete the attribute, and the return the application instance

**Reply by jozyonline** [reply]
this approach works really well and efficiently thanks a lot.

**Reply by dickmartinez** [reply]
you rule dude!, thanks a lot

**Reply by MuhRiswan** [reply]
thanks sir its work on Inertia React!

**Reply by mhdasad** [reply]
really appreciate this, thank you sir 👌
